### PR TITLE
GH1139 Series.rename inplace

### LINKED ISSUE #1139: Series().rename(..., inplace=True) can return a Series if the renamer is hashable
Series.rename inplace can actually return Self and not None see https://github.com/pandas-dev/pandas/issues/60942
The current typehinting would force it to be None but at runtime this does not seem to stand.

```python
>>> import pandas as pd
>>> pd.Series([1, 2, 3]).rename("A", inplace=True)
0    1
1    2
2    3
Name: A, dtype: int64
```

### PR DESCRIPTION
**Pull request by loicdiridollou**
- [x] Closes #1139 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

Added some test framework migration as this was how I originally found the issue with `rename`

### CODE REVIEW

File: pandas-stubs/core/series.pyi [hunk 1]
```diff
@@ -1041,9 +1041,20 @@ class Series(IndexOpsMixin[S1], NDFrame):
...
+        axis: Axis | None = ...,
+        copy: bool = ...,
+        inplace: Literal[True],
+        level: Level | None = ...,
+        errors: IgnoreRaise = ...,
+    ) -> Self: ...
+    @overload
+    def rename(  # type: ignore[overload-cannot-match]
+        self,
+        index: Renamer | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:1057]: reverse the order of these two, and then make the one with `Hashable` be `index: Hashable` so that it has to be specified.  That will change the ignore-related comments as well.

File: tests/test_series.py [hunk 2]
```diff
@@ -1137,7 +1138,6 @@ def test_types_set_flags() -> None:
 
 def test_types_getitem() -> None:
     s = pd.Series({"key": [0, 1, 2, 3]})
-    key: list[int] = s["key"]
```
Comment (Dr-Irv) [inline tests/test_series.py:1140]: I think this was supposed to be a test.  Should be 
```python
    check(assert_type(s["key"], Any), list)
```

File: tests/test_series.py [hunk 3]
```diff
@@ -1212,7 +1228,10 @@ def add1(x: int) -> int:
     check(assert_type(s5, "pd.Series[int]"), pd.Series, np.integer)
     # inplace
     # TODO fix issue with inplace=True returning a Series, cf pandas #60942
-    s6: None = pd.Series([1, 2, 3]).rename("A", inplace=True)
+    check(
+        assert_type(pd.Series([1, 2, 3]).rename("A", inplace=True), "pd.Series[int]"),
+        pd.Series,
+    )
```
Comment (Dr-Irv) [inline tests/test_series.py:1234]: You should add a test for specifying `inplace=True` with `index=None` being explicit, as well as a test for specifying `inplace=True` with `index` being a dict.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

**Comment by Dr-Irv** [comment]
@loicdiridollou let me know when you think you'll get to this.  I'd like to do a new release soon, and want to know if I should wait for this change or not.

**Comment by loicdiridollou** [comment]
I was planning to take care of it in the evening but if you need to make a
release feel free to go ahead !3;943 that.

File: tests/test_series.py [hunk 4]
```diff
@@ -70,6 +70,7 @@
     TimedeltaSeries: TypeAlias = pd.Series
     TimestampSeries: TypeAlias = pd.Series
     OffsetSeries: TypeAlias = pd.Series
+    ExtensionArray: TypeAlias = np.ndarray
```
[the remark this answers is not in the document]
Comment (loicdiridollou) [reply tests/test_series.py:73]: It is needed for the tests I had converted since it is internal to the stubs

[the remark this answers is not in the document]
Comment (Dr-Irv) [reply tests/test_series.py:73]: Shouldn't be.  It's imported at the top of the file on line 30.

File: tests/test_series.py [hunk 5]
```diff
@@ -1211,8 +1228,26 @@ def add1(x: int) -> int:
...
+        assert_type(
+            pd.Series([1, 2, 3]).rename(index=None, inplace=True), "pd.Series[int]"
+        ),
+        pd.Series,
+        np.integer,
+    )
+    # TODO this should not raise an error, the lambda is matched with Hashable
+    # check(
+    #     assert_type(pd.Series([1, 2, 3]).rename(lambda x: x**2, inplace=True), None), type(None)
+    # )
```
[the remark this answers is not in the document]
Comment (loicdiridollou) [reply tests/test_series.py:1250]: If I do so mypy seems to match the return of the expression with Any instead of None, let me see how I can address

[the remark this answers is not in the document]
Comment (loicdiridollou) [reply tests/test_series.py]: I have pushed for you to take a look but not sure how to address, this is a mypy issue and not pyright, pyright is happy with the current setup and does not flag it.

File: tests/test_series.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply tests/test_series.py:73]: I checked - you can remove this.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
I got it to work.  The issue is that `Callable` and `Hashable` are overlapping and `mypy` is having trouble with that.  So instead of using `Hashable`, I used `Scalar | tuple[Hashable, ...]` which corresponds to the docs anyway.

So my review contains all the changes (as suggestions) that you should need to make.

File: tests/test_series.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (loicdiridollou) [reply tests/test_series.py:73]: It gives me error when I remove it as it creates a lot of warnings for line 1187 and below

[the remark this answers is not in the document]
Comment (Dr-Irv) [reply tests/test_series.py:73]: What kinds of errors?  I removed it and it passed my local tests.  

Make sure your `pyright` and `mypy` are up to date via `poetry update`

[the remark this answers is not in the document]
Comment (loicdiridollou) [reply tests/test_series.py:73]: They are up to date, the issue comes when the pre-commit triggers and deleting the one below will also delete the one on top and then the tests referencing `ExtensionArray` are now unable to find what it is 

```
--- a/tests/test_series.py
+++ b/tests/test_series.py
@@ -27,7 +27,6 @@ import numpy as np
 import pandas as pd
 from pandas._testing import ensure_clean
 from pandas.api.extensions import (
-    ExtensionArray,
     ExtensionDtype,
 )
 from pandas.core.arrays.datetimes import DatetimeArray
@@ -70,7 +69,6 @@ else:
     TimedeltaSeries: TypeAlias = pd.Series
     TimestampSeries: TypeAlias = pd.Series
     OffsetSeries: TypeAlias = pd.Series
-    ExtensionArray: TypeAlias = np.ndarray

 if TYPE_CHECKING:
     from pandas._typing import (
```

[the remark this answers is not in the document]
Comment (loicdiridollou) [reply tests/test_series.py:73]: I deleted the one at the bottom (line 73) and then I run the full process `poetry run poe test_all` and at the end the pre-commit for ruff.

[the remark this answers is not in the document]
Comment (Dr-Irv) [reply tests/test_series.py:73]: OK, the problem is that you have `"ExtensionArray | np.ndarray"` in the test.  Change it to `Union[ExtensionArray, np.ndarray]` and then it should be fine.

**Review by Dr-Irv — APPROVED** [review]
thanks @loicdiridollou
